contracts/api3-server-v1/proxies/CompositeApi3ReaderProxyV1.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import "./interfaces/ICompositeApi3ReaderProxyV1.sol";
+import "../../interfaces/IApi3ReaderProxy.sol";
+
+import "hardhat/console.sol";
+
+/// @title An immutable proxy contract that is used to read a composite data feed
+/// by multiplying two IApi3ReaderProxy data feeds
+/// @dev This contract implements the AggregatorV2V3Interface to be compatible with
+/// Chainlink aggregators. This allows the contract to be used as a drop-in
+/// replacement for Chainlink aggregators in existing dApps.
+/// Refer to https://github.com/api3dao/migrate-from-chainlink-to-api3 for more
+/// information about the Chainlink interface implementation.
+contract CompositeApi3ReaderProxyV1 is ICompositeApi3ReaderProxyV1 {
+    /// @notice First IApi3ReaderProxy contract address
+    address public immutable override proxy1;
+
+    /// @notice Second IApi3ReaderProxy contract address
+    address public immutable override proxy2;
+
+    /// @notice Decimals of the rate composition
+    uint8 public immutable override decimals;
+
+    uint256 private immutable baseUnit;
+
+    /// @dev Proxy addresses must be different and decimals must be a value
+    /// between 1 and 18 inclusive
+    /// @param proxy1_ First IApi3ReaderProxy contract address
+    /// @param proxy2_ Second IApi3ReaderProxy contract address
+    /// @param decimals_ Decimals of the rate composition
+    constructor(address proxy1_, address proxy2_, uint8 decimals_) {
+        if (proxy1_ == address(0) || proxy2_ == address(0)) {
+            revert ZeroProxyAddress();
+        }
+        if (proxy1_ == proxy2_) {
+            revert SameProxyAddress();
+        }
+        if (decimals_ == 0 || decimals_ > 18) {
+            revert InvalidDecimals();
+        }
+        proxy1 = proxy1_;
+        proxy2 = proxy2_;
+        decimals = decimals_;
+        baseUnit = 10 ** decimals_;
+    }
+
+    /// @notice Returns the current value and timestamp of the rate composition
+    /// between two API3 data feeds associated with the two IApi3ReaderProxy
+    /// contracts
+    /// @dev The value is calculated by multiplying the values returned by each
+    /// proxy and this may overflow if the multiplication results in a value
+    /// unsuitable for an int256. The timestamp is the current block timestamp
+    /// @return value Value of the rate composition
+    /// @return timestamp Timestamp of the rate composition
+    function read()
+        public
+        view
+        override
+        returns (int224 value, uint32 timestamp)
+    {
+        (int224 value1, ) = IApi3ReaderProxy(proxy1).read();
```

Comment:
If we go with IApi3ReaderProxy requirement here I think it makes sense to hardcode 18 for decimals in the implementation. I like that the implementation is very clean this way. If we want to be generic, we can expect `latestAnswer` as Burak pointed out, but I prefer the current implementation.